parsing/location.ml

```diff
@@ -292,90 +292,234 @@ let highlight_terminfo lb ppf locs =
-      in
-      loop 0 0 0;
-      !line_start, !line_end, !line_start_pos, !line_end_pos
+    let leftmost =
+      List.sort (fun l l' -> compare l.loc_start.pos_cnum l'.loc_start.pos_cnum)
+        locs
+      |> List.hd |> fun l -> l.loc_start
+    in
+    let rightmost =
+      List.sort (fun l l' -> compare l'.loc_end.pos_cnum l.loc_end.pos_cnum)
+        locs
+      |> List.hd |> fun l -> l.loc_end
+    in
+    let is_start pos = List.exists (fun l -> l.loc_start.pos_cnum = pos) locs in
+    let is_end pos = List.exists (fun l -> l.loc_end.pos_cnum = pos) locs in
+    let try_get_line_number start_cnum end_cnum =
+      try
+        List.map (fun l -> [l.loc_start; l.loc_end]) locs
+        |> List.flatten
+        |> List.find (fun p -> start_cnum <= p.pos_cnum &&
+                               p.pos_cnum <= end_cnum)
+        |> fun pos_in_line -> Some pos_in_line.pos_lnum
+      with Not_found -> None
+    in
+    let lines =
+      get_lines ~start_pos:leftmost ~end_pos:rightmost
+      |> List.map (fun (line, s) ->
+        line, try_get_line_number s (s + String.length line), s)
+    in
+    let lines_nb_maxlen =
+      List.fold_left (fun acc (_, line_nb_opt, _) ->
+        match line_nb_opt with
+        | Some line_nb -> max acc (String.length (string_of_int line_nb))
+        | None -> acc
+      ) 0 lines
+    in
+    let prefix ~sep ppf lnum_opt =
+      Format.fprintf ppf "%*s %s "
+        lines_nb_maxlen
+        (Misc.Stdlib.Option.value_default string_of_int ~default:"" lnum_opt)
+        (if sep then "|" else " ")
     in
     Format.fprintf ppf "@[<v>";
-    (* Print the input, highlighting the locations.
-       Indent by two spaces. *)
-    Format.fprintf ppf "  @[<v>";
-    if line_start = line_end then begin
-      (* single-line error *)
-      for pos = line_start_pos to line_end_pos - 1 do
-        Format.pp_print_char ppf (read_char pos)
-      done;
-      Format.fprintf ppf "@,";
-      for pos = line_start_pos to rightmost_end - 1 do
-        if is_highlighted pos then Format.pp_print_char ppf '^'
-        else Format.pp_print_char ppf ' '
-      done
-    end else begin
-      (* multi-line error *)
-      for pos = line_start_pos to line_end_pos - 1 do
-        let c = read_char pos in
-        if c = '\n' then Format.fprintf ppf "@,"
-        else if is_highlighted pos then Format.pp_print_char ppf c
-        else Format.pp_print_char ppf '.'
-      done
+    begin match lines with
+    | [] | [("", _, _)] -> ()
+    | [(line, line_nb_opt, line_start_cnum)] ->
+        (* Single-line error *)
+        Format.fprintf ppf "%a%s@," (prefix ~sep:true) line_nb_opt line;
+        Format.fprintf ppf "%a" (prefix ~sep:false) None;
+        let highlighted = ref false in
+        for pos = line_start_cnum to rightmost.pos_cnum - 1 do
+          if is_end pos then begin
+            highlighted := false;
+            Format.fprintf ppf "@}"
+          end;
+          if is_start pos then begin
```

Comment:
I'm again uneasy with the linear lookup here, giving quadratic behavior. We could have a set of start positions.